game-core/src/main/java/games/strategy/engine/framework/startup/ui/posted/game/pbem/EmailSenderEditor.java

```diff
@@ -177,6 +187,82 @@ public void viewModelChanged(final EmailSenderEditorViewModel viewModel) {
           toggleFieldVisibility();
           emailProviderSelectionBox.setSelectedItem(viewModel.getSelectedProvider());
 
+          if (viewModel.isEmailProviderDisabled()) {
```

Comment:
Design nit, the view model attempts to remove all logic from the view. This can be done by creating a method to fetch the help message directly from the view model, or secondarily to create a method to tell the view which help text to render. This way the view asks the view model which help text to render as opposed to the view calculating which help text to render.

EG:

```
 helpMessage.setText(viewModel.getEmailSenderHelpText());
```

Or:
```
switch(viewModel.getHelpTextToDisplay()) {
  case NO_EMAIL_PROVIDER_SELECTED:
      helpMessage.setText( ...);
      break;
   case GMAIL_EMAIL_PROVIDER:
     :
     : 
}
```

The above is different as currently view model is interpreting the meaning of `isEmailProviderDisabled` rather than being told directly what to render.

For a viewmodel pattern, the first variant of getting the help message is probably closest to the pattern. The switch statement with a dedicated flag condition would work too.

Reply:
Out of scope. Feel free to follow on adjusting it if you want.

Reply:
I really don't think it is out of scope:
- adds something that we really need to do that should be done now.
- very much breaks pattern and consistency
- misses updates to test cases, adding more stuff to do later.

ViewModel is a pretty powerful pattern: https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93viewmodel, it was with quite some effort it was introduced to line up for the migration to javaFx where we would not need to update the views.

Reply:
The update to get back to pattern is not very time consuming and changes the accessor methods added to the view model. The testing should be pretty quick as well, but we are talking 20 minutes of work. Opening up the view model as was done is breaking to the design pattern.

Reply:
I really don't think there is much difference between your switch example and what is implemented. Either way the view has to interpret what to do based on either some form of enumeration/key like your example (getHelpTextToDisplay) or interpret whether to display certain text based on a few booleans. In the end, you always have to draw the line somewhere and try to keep the "business logic" out of the view but where that line is drawn can always be debated to some extent which can be shown by your 2 examples which take it to different degrees.

Reply:
It's not that much different, perhaps there should be no objection then to fit squarely into the pattern?

The example we have here is really similar to: https://martinfowler.com/eaaDev/SeparatedPresentation.html

The 'boolean' parameter rather than an integer is less obvious of a parallel, but it's still state  the view interpreting state (logic) rather than being told what to do. In part, the pattern uses dedicated flags and switches to tell the view what to do. The same could be said about the different boolean parameters that control the various buttons. Many have the same logic condition behind them, yet the view model exposes those attributes as different boolean methods. In this way the view is totally unaware of the backend state, nor does it know that any state is coupled to anything else.

There are also two or three other helper methods for the 'isGmail' etc that could drop away if the view model were to be passed the text to be rendered.

Reply:
If the view were to call: `getHelpTextToDisplay`, then it would have nothing to interpret. The switch is similar, but it's still different as one is inferring state, as opposed to being told what state to be in.

Reply:
Like notice we could have done the same with:
```
    emailProviderSelectionBox.setSelectedItem(viewModel.getSelectedProvider());
```
That could have been a very similar 'if/else' block as well. It's not though to prevent the view from having to evaluate boolean conditions.  Let's do that consistently please.

Reply:
> Either way the view has to interpret what to do based on either some form of enumeration/key like your example (getHelpTextToDisplay) or interpret whether to display certain text based on a few booleans

I don't think that fits. The enumeration/key would be dedicated and used in only one place in the view, and the enumeration would be provided to the view and not used internally by the view model. The enumeration/key would be a computed value, an output, it would not be an input used elsewhere. The output would be dedicated for the view component. In this example we are moving logic into the view.

It would also never be multiple booleans, the only boolean would be a single one. We really should have a string here to think about it, the state of the component is essentially a string. A view-model is the state of each component translated into a primitive. A button enabled = a single boolean, the text of an input field is a string. For a dynamic help text, it is a string. 

Think about it in this way, the view is pure presentation. We should be able to simply swap out the swing view for javafx and only need to update the component logic, the rest of the state logic should be able to stay the same. With javafx migration, we should be able to re-use exactly the view-model objects. It's argument for the help text to live in the view-model.

Reply:
Piling on slightly, the switch and if statement are actually quite different.
The switch is iterating over a single value, it's a single value that controls the state of the UI. It is not the UI checking multiple parts of the backing state. 

A method to get the value of the text, is also a scalar. At the end of the day, the backend should give the front-end a single scalar to tell it how to render. Whether that scalar is a flag, or if it's the actual text, it's still a single value. It is not the view querying the view-model and doing logic.